### PR TITLE
[react] Improve type-checking of state updater functions

### PR DESCRIPTION
**Pull request by EthanStandel**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [The code that validates whether a state should be executed acquire the value or if it is the value.](https://github.com/facebook/react/blob/b72ed698fb5f15e0c33c5017422b8428b01ff12a/packages/react-reconciler/src/ReactFiberHooks.js#L1040)

I would also like to mention that this package seems to be heavily derived from Flow types in the source code of React. The types in React __do not properly reflect my change__, but the behavior does. If the component I built for this test ran, it would throw an error and the runtime state would be out of sync with the TS type. I also don't know if Flow supports conditional types anyways, so I don't know if I can even suggest this type in a PR to React itself.

### CODE REVIEW

**Comment by EthanStandel** [comment]
As it turns out, because `boolean` in TypeScript is, in far too many contexts, broken down to `true | false`, my change has had the unintended consequence of breaking toggling `useState` cases, as `setState` will only accept `value: boolean | ((prevState: false) => false) | ((prevState: true) => true)`... so you can't give it `x => !x`. This is likely unfortunate dealbreaker on this change as obviously this edge case doesn't beat out being able to have a state toggle.

I'll leave this open for a bit and try to see if I can find another solution, but I'm not sure one exists for this case.

**Review by eps1lon — CHANGES_REQUESTED** [review]
Thanks for doing this. This looks very promising! Could you rebase onto master since I just merged some more extensive tests in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64882 to help understand how this change affects `useState`?

I think once you rebase you can drop the added tests and just change the tests added in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64882 accordingly.

**Comment by EthanStandel** [comment]
Thanks for the review @eps1lon! Sorry I've taken so long to update, I' Just updated and switched your noted `Undesired` behavior to be `@ts-expect-error`. It passes locally when I run `npm test react` locally, however, looks like another package is having a somewhat unrelated failure but it depends on React. It appears that `wordpress_components` is specifically failing in TypeScript 5.1. The place where it's failing uses a function as JSX children and expects arguments of that function to be implicitly typed which I think may be an issue with the changes to JSX in 5.1

## The error

```
=== ERRORS ===



Error in wordpress__components
Error: /home/vsts/work/1/s/types/wordpress__components/wordpress__components-tests.tsx:291:9
ERROR: 291:9  expect  TypeScript@5.1 compile error: 
Binding element 'onClose' implicitly has an 'any' type.

    at testTypesVersion (/home/vsts/work/1/s/node_modules/@definitelytyped/dtslint/src/index.ts:235:11)
    at runTests (/home/vsts/work/1/s/node_modules/@definitelytyped/dtslint/src/index.ts:176:5)
```

## The associated code: 

```tsx
<C.DropdownMenu icon="move" label="Select a direction">
    {({ onClose }: { onClose: () => void; }) => (
        <div>
            <button onClick={onClose}>Click me</button>
        </div>
    )}
</C.DropdownMenu>;
```

Now, to ensure this at least gets seen, I'm going to attempt to "fix" the "error" by _explicitly_ defining the argument type but from what I'm seeing the implicit type should work. I also think this has little to nothing to do with the actual change I'm trying to make.

File: types/wordpress__components/wordpress__components-tests.tsx [hunk 1]
```diff
@@ -288,7 +288,7 @@ const buttonGroupRef = createRef<HTMLDivElement>();
 // dropdown-menu
 //
 <C.DropdownMenu icon="move" label="Select a direction">
-    {({ onClose }) => (
+    {({ onClose }: { onClose: () => void; }) => (
```
Comment (EthanStandel) [inline types/wordpress__components/wordpress__components-tests.tsx:291]: Ftr, I don't like that I had to do this and this doesn't seem right to me.

[Here's the test failure that inspired this change.](https://github.com/DefinitelyTyped/DefinitelyTyped/actions/runs/4770903812/jobs/8482493345#step:6:1894)

Comment (eps1lon) [reply types/wordpress__components/wordpress__components-tests.tsx:291]: You can revert that for now. The [issue is being tracked on the TS repo](https://github.com/microsoft/TypeScript/issues/53941) and I can always merge DT PRs with red CI (packages will be published anyway)

**Review by eps1lon — COMMENTED** [review]
I wouldn't worry about Flow types here. We want to match the documented behavior of React not mirror Flow types (which neither could do since Flow and TS choose different typing tradeoffs).

I'll check this patch real quick on my work codebase and hope this integrates cleanly

**Review by eps1lon — CHANGES_REQUESTED** [review]
Just leaving the request changes for clarity until https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64567#discussion_r1174311960 is resolved.

**Review by eps1lon — COMMENTED** [review]
Oh and we have a fork for TS 5.0 now. Can you port these changes to the ts5.0 folder as well?

**Review by eps1lon — CHANGES_REQUESTED** [review]
Unfortunately, the proposed implementation doesn't work if the state is generic (even if that generic is constrained to not be a function) as far as I can tell.

This should pass type-checking but doesn't with the new implementation.

```tsx
import * as React from 'react'

function useMyState<S extends string>(value: S) {
    const [, setValue] = React.useState(value)
    setValue(value)
}
```

[Playground Link using published types](https://www.typescriptlang.org/play?ts=5.0.4#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQFA0ECuAdvsBE3A6lgLICeAZRjIYWADwC4WAB6imAE3SoYUYEwDmAPgAUAN2QAbBlgBccAQEo4Abxpx7cXO2VwA2gBo43GADVDxgF04AF5MKhgAOi4sIREsPX8sCzsHbz8jeP0M5IBfIA)

Could you add a test for that to confirm?

**Comment by EthanStandel** [comment]
@eps1lon So I've updated the PR with the requested updates, but the case from your last comment is , but I'm very somewhat of an unfixable issue but I'm glad you found it. It took me a minute to realize why this is happening but after I think about it, the error makes sense even though it's technically wrong. Even though this type is impossible to create in JavaScript, TypeScript considers it perfectly valid.

```ts
type Impossible = string & (() => void);
```

So from TypeScript's perspective in your example, just because `S extends string` doesn't mean it couldn't _also_ extend a function. So TypeScript's perspective would be that with the change in this PR, your example _can't be valid_ and just for safety's sake you must use a functional setter. So this works.

```ts
import * as React from 'react'

function useMyState<S>(value: S) {
    const [, setValue] = React.useState(value)
    setValue(() => value)
}
```

But I could see how this could create serious ripples in the validity of types that are either definitely valid or valid 99% of the time in a lot of libraries, all to support functional state which is itself an edge case. So if this is a deal-breaker for this PR, I would totally understand. If you think it's a dealbreaker, you can close it or you can confirm so to me and I can close it.

**Comment by eps1lon** [comment]
> So from TypeScript's perspective in your example, just because S extends string doesn't mean it couldn't also extend a function.

I probably misunderstand the semantics that you're talking about but from a user perspective, this doesn't hold true. `() => number` would not be assignable to `T` iff `T` is constrained to `extends string`.

```ts
import * as React from 'react'

function useMyState<S extends string>(value: S) {
    const [, setValue] = React.useState(value)
    setValue(() => value)
}

// This does not typecheck as expected. 
// So `useMyState` will only ever see strings.
// But the new implementation of `useState` would not accept `setValue(value)`
useMyState<() => number>(() => 1);
```

[Playground Link](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQFA0ECuAdvsBE3A6lgLICeAZRjIYWADwC4WAB6imAE3SoYUYEwDmAPgAUAN2QAbBlgBccAQEo4Abxpx7cXO2VwA2gBo43GADVDxgF04AF5MKhgAOi4sIREsPX8sCzsHbz8jeO0rYM04fQzkgF86AHoSuAAVAAtgdHkILHQmCHgYPjAsXCrOgGsUdBkO-Cx5CLgaMvMIOAADaP5Y0Rm4AHdgAwM4dgM+KV0sKC8sLC8VNXVUGnnBYVExLJDcpgYQACMDnQecuABGCwBuIA)

I think this is a TypeScript issue and we're probably not the first ones encountering it. Would love to know if there has been prior discussions around this somewhere else.

**Comment by eps1lon** [comment]
Hm we need a way to teach this since it's not obvious why previously fine patterns would now be rejected. For the 2.7M TS LoC codebase we had one legitimate case being flagged, two cases that  are bulletproof (e.g. `T extends string`), and 10 potentially unsafe cases that are not actually unsafe (e.g. just `useState<T>` but `T` was always used for primitives like string number etc). 

That's obviously not sufficient to block this but I felt way more comfortable if it would be reversed: 95% true positives and 5% false positives instead of 95% false positives and 5% true positives I saw.

What would we recommend people to write in the `useState<T extends string>` case? How would I write `setState(someString)` without outright `@ts-expect-error`?

**Comment by EthanStandel** [comment]
@eps1lon On the ratio of false-positives to true-positives, I'm not surprised. I only came across this case because I was building a small PoC hook library and was trying to consider all the possible cases myself. But in actual applications, I've only needed functional state 1 time. So it's definitely a rare edge case for what this will catch and protect people from.

So I don't think it's unreasonable to look at this change and say that it's not worth it for the potential fallout. This might be the case of types that are cautious to a fault and the kind of thing that just frustrates TypeScript holdouts. If you want to make that call, I'm fine with closing this out for that perspective.

But I do think it's correct and I don't think it requires `@ts-expect-error`/`@ts-ignore`. If you want to take in a generic value and be able to set that value reliably, then I do think the truly safe way to do that would be to use a state setting function rather than passing the value directly, so changing any generic instances of `setState(value)` to `setState(() => value)`. And even if you shouldn't _need_ to do it (like if your generic `extends string`), there's still no harm. And if you wanted to expose the functional setter from your custom hook, you could also have the component accept `SetStateAction<S>`, e.g.

```ts
function useMyState<S>(updatedValue: React.SetStateAction<S | undefined>) {
    const [, setValue] = React.useState<S | undefined>()
    setValue(updatedValue)
}

useMyState("hello"); // S is <string>
useMyState(() => "hello"); // S is <string>
useMyState(() => () => "hello"); // S is <() => string>
```

Obviously, this case makes it so that you can't use the initial value as a setter as well, but to me that might be the moral of this topic: initializing state & setting new state aren't actually the same operation.

It's too bad React doesn't have more of a direct integration with it's TypeScript types. It would be nice to just be able to document this case in the React docs as a potential footgun of using functional state, while expressing that it wasn't worth the risk of fixing.

**Pull request closed without merging**